src/rewrite/vct/rewrite/lang/LangCPPToCol.scala

```diff
@@ -728,6 +1132,183 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre]) extends L
+      CPP.unwrappedType(base.t) match {
+        case SYCLTAccessor(_, 2) => ArraySubscript[Post](
+          Deref[Post](currentThis.get, accessor.instanceField.ref)(new SYCLAccessorDerefBlame(accessor.instanceField)),
+          ProcedureInvocation[Post](linearizedIndexProc, linearizeArgs, Nil, Nil, Nil, Nil)(SYCLAccessorArraySubscriptLinearizeInvocationBlame(sub, base, Seq(indexX, indexY)))
+        )(SYCLAccessorArraySubscriptErrorBlame(sub))
+        case t: SYCLTAccessor[Pre] => throw SYCLWrongNumberOfSubscriptsForAccessor(sub, 2, t.dimCount)
+      }
+    case AmbiguousSubscript(AmbiguousSubscript(AmbiguousSubscript(base: CPPLocal[Pre], indexX), indexY), indexZ) if CPP.unwrappedType(base.t).isInstanceOf[SYCLTAccessor[Pre]] =>
+      implicit val o: Origin = sub.o
+      val accessor = currentAccessorSubstitutions(base.ref.get)
+      val linearizeArgs = Seq(rw.dispatch(indexX), rw.dispatch(indexY), rw.dispatch(indexZ),
+        Deref[Post](currentThis.get, accessor.rangeIndexFields(0).ref)(new SYCLAccessorDimensionDerefBlame(accessor.rangeIndexFields(0))),
+        Deref[Post](currentThis.get, accessor.rangeIndexFields(1).ref)(new SYCLAccessorDimensionDerefBlame(accessor.rangeIndexFields(1))),
+        Deref[Post](currentThis.get, accessor.rangeIndexFields(2).ref)(new SYCLAccessorDimensionDerefBlame(accessor.rangeIndexFields(2)))
+      )
+      val linearizedIndexProc = getSYCLHeaderMethodRef("sycl::accessor::linearize_3_indices", linearizeArgs)
+      CPP.unwrappedType(base.t) match {
+        case SYCLTAccessor(_, 3) => ArraySubscript[Post](
+          Deref[Post](currentThis.get, accessor.instanceField.ref)(new SYCLAccessorDerefBlame(accessor.instanceField)),
+          ProcedureInvocation[Post](linearizedIndexProc, linearizeArgs, Nil, Nil, Nil, Nil)(SYCLAccessorArraySubscriptLinearizeInvocationBlame(sub, base, Seq(indexX, indexY, indexZ)))
+        )(SYCLAccessorArraySubscriptErrorBlame(sub))
+        case t: SYCLTAccessor[Pre] => throw SYCLWrongNumberOfSubscriptsForAccessor(sub, 3, t.dimCount)
+      }
+    case _ => rw.rewriteDefault(sub)
+  }
+
+  def rewriteScope(scope: CPPScope[Pre]): Statement[Post] = {
+    syclBufferSuccessor.push(mutable.Map.empty)
+    syclRunningKernelsSuccessor.push(mutable.Map.empty)
+
+    val result: Scope[Post] = rw.dispatch(Scope[Pre](scope.locals, scope.body)(scope.o)).asInstanceOf[Scope[Post]]
+
+    implicit val o: Origin = scope.o
+
+    // Wait for SYCL kernels to finish executing
+    var additionalStatements: Seq[Statement[Post]] = syclRunningKernelsSuccessor.pop().map(tuple => syclKernelTermination(tuple._1, tuple._2)).toSeq
+
+
+    // Destroy all buffers and copy their data back to host
+    additionalStatements =  additionalStatements ++ syclBufferSuccessor.pop().map(tuple => destroySYCLBuffer(tuple._2, scope)).toSeq
+
+    Scope[Post](result.locals, Block[Post](result.body +: additionalStatements)(result.body.o))(scope.o)
+  }
```

Comment:
I'm guessing this normally happens with the RAII pattern via destructors? Do you think it is infeasible to support destructors generally? I'm a bit worried about `CPPScope` because of two things:

* The scoping part of a `CPPScope` is not meaningfully different from `Scope`, so changes to `Scope` logic should likely be reflected in `CPPScope`?
* I doubt that destructors are always related to the scope in the way I read it here, surely you can move the values or delete them earlier somehow (but I'm far from a c++ expert).

Idea: continue to use scope, and introduce a node `CppLifetimeScope(body: Statement)` that analyses what destructors to call (or emulate for sycl stuff), this at least decouples the concerns of what variables you're allowed to name and what resources need to be cleaned up.

Reply:
> I'm guessing this normally happens with the RAII pattern via destructors?

Yes, indeed. For SYCL buffers and event objects, some extra actions are performed in their destructors, other than releasing the associated memory, such as copying data back and waiting on kernels to finish running. I believe destructors can also be called in other ways than the assigned variable going out of scope (but I have also not worked with destructors in C++ before), but for the SYCL implementation it was enough to only support 'invoking' them when going out of scope.

> Idea: continue to use scope, and introduce a node CppLifetimeScope(body: Statement) that analyses what destructors to call (or emulate for sycl stuff), this at least decouples the concerns of what variables you're allowed to name and what resources need to be cleaned up.

I implemented this now, it is indeed a better solution.